### PR TITLE
Add CORS headers

### PR DESCRIPTION
**Pull request by ger-benjamin**
In response to #1706

### CODE REVIEW

File: c2cgeoportal/views/entry.py [hunk 1]
```diff
@@ -876,7 +876,7 @@ def get_cgxp_index_vars(self, templates_params={}):
         except:  # pragma: no cover
             d["mobile_url"] = None
 
-        set_common_headers(self.request, "cgxp_index", NO_CACHE)
+        set_common_headers(self.request, "cgxp_index", NO_CACHE, add_cors=True)
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no CORS on index page ...

File: c2cgeoportal/views/entry.py [hunk 2]
```diff
@@ -932,14 +932,14 @@ def get_cgxp_viewer_vars(self):
             d["permalink_themes"] = json.dumps(permalink_themes.split(","))
 
         set_common_headers(
-            self.request, "cgxp_viewer", PRIVATE_CACHE,
+            self.request, "cgxp_viewer", PRIVATE_CACHE, add_cors=True,
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no cors on viewer.js page

File: c2cgeoportal/views/entry.py [hunk 3]
```diff
@@ -962,7 +962,9 @@ def mobile(self):
         """
         View callable for the mobile application's index.html file.
         """
-        set_common_headers(self.request, "sencha_index", NO_CACHE)
+        set_common_headers(
+            self.request, "sencha_index", NO_CACHE, add_cors=True
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no cors on mobile page

File: c2cgeoportal/views/entry.py [hunk 4]
```diff
@@ -1080,7 +1082,7 @@ def mobileconfig(self):
                 })
 
         set_common_headers(
-            self.request, "sencha_config", PRIVATE_CACHE,
+            self.request, "sencha_config", PRIVATE_CACHE, add_cors=True,
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no cors on mobile page...

File: c2cgeoportal/views/entry.py [hunk 5]
```diff
@@ -1141,7 +1143,7 @@ def xapijs(self):
 
     @view_config(route_name="apihelp", renderer="api/apihelp.html")
     def apihelp(self):
-        set_common_headers(self.request, "apihelp", NO_CACHE)
+        set_common_headers(self.request, "apihelp", NO_CACHE, add_cors=True)
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no cors on api help pages (also xapi)

File: c2cgeoportal/views/entry.py [hunk 6]
```diff
@@ -1219,7 +1221,9 @@ def loginform403(self):
         if self.request.authenticated_userid:
             return HTTPForbidden()  # pragma: no cover
 
-        set_common_headers(self.request, "loginform403", NO_CACHE)
+        set_common_headers(
+            self.request, "loginform403", NO_CACHE, add_cors=True
+        )
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: no cors on html pages, also next one

File: c2cgeoportal/views/exportcsv.py [hunk 7]
```diff
@@ -62,6 +62,6 @@ def echo(request):
         name.replace(" ", "_"), csv_extension
     )
     return set_common_headers(
-        request, "csvecho", NO_CACHE,
+        request, "csvecho", NO_CACHE, add_cors=True,
```
Comment (sbrunner) [inline c2cgeoportal/views/exportcsv.py]: echo like cors is not useful ..

**Comment by ger-benjamin** [comment]
(please, wait before review)

**Comment by sbrunner** [comment]
Looks good for me :-)

**Comment by elemoine** [comment]
How did you decide where to use CORS and where not to use CORS?

**Comment by ger-benjamin** [comment]
I always use CORE except:
- In case of the service return privates informations.
- If this is "html services", or services used near-obviously only by the same server.

**Comment by elemoine** [comment]
I am a little unsure (and nervous) about enabling CORS for so many things. Are we clear on the implications? Are we sure we're not opening breaches?

File: c2cgeoportal/views/entry.py [hunk 8]
```diff
@@ -1276,11 +1277,12 @@ def logout(self):
         return set_common_headers(
             self.request, "logout", NO_CACHE,
             response=Response("true", headers=headers),
+            add_cors=True
         )
 
     @view_config(route_name="loginchange", renderer="json")
     def loginchange(self):
-        set_common_headers(self.request, "loginchange", NO_CACHE)
+        set_common_headers(self.request, "loginchange", NO_CACHE, add_cors=True)
```
Comment (elemoine) [inline c2cgeoportal/views/entry.py]: This means a CSRF could now be used to change the user's password. Same thing for loginresetpassword.

**Comment by ger-benjamin** [comment]
I understand and I'm not competent in the domain of the security. But I really don't see problems to open services like "themes", "fulltextsearch" or even "cgxp_index".

But yeah, I'm a little bit nervous too with the "login-related" services...

File: c2cgeoportal/views/entry.py (hunk 8 above)
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: I don't understand the relation between CORS and CSRF to change the password eventually about reading the result but not change the password.
For me we already have an issue...
And if we don't do that we cant do an ngeo example to demo server ... (for the concerned functionalities)
I am wrong?

Comment (elemoine) [inline c2cgeoportal/views/entry.py]: Maybe CSRF is not the correct term to use. I am not sure. BUT I do think enabling CORS creates a breach here. This makes it possible to create a page that can change the user's password by sending an Ajax request to GeoMapFish's loginchange web service. No?

**Comment by elemoine** [comment]
> But I really don't see problems to open services like "themes", "fulltextsearch" or even "cgxp_index".

I am not a security expert either. But I think enabling CORS makes it possible for an attacker to make cross-site Ajax requests to the "themes" or "fulltextsearch" web services with the user's credentials, and therefore get access to unauthorized data.

File: c2cgeoportal/views/entry.py (hunk 8 above)
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: For me the break is already open with or without CORS...

**Comment by sbrunner** [comment]
As I understand this is effectively correct ...

File: c2cgeoportal/views/entry.py (hunk 8 above)
Comment (elemoine) [inline c2cgeoportal/views/entry.py]: Yep, you may be right.

**Comment by sbrunner** [comment]
A solution can be that we list a set of domains that is allows to read the services ...
To fill the `Access-Control-Allow-Origin` with the list instance of with a `*`.

**Comment by elemoine** [comment]
> A solution can be that we list a set of domains that is allows to read the services ...
> To fill the Access-Control-Allow-Origin with the list instance of with a *.

Yes exactly. We should be more restrictive.

**Comment by ger-benjamin** [comment]
Ok, thanks @elemoine, @sbrunner I'll work on that.

**Comment by ger-benjamin** [comment]
Something like my latest commit ? If the principle suits you, I'll add some documentation here: http://docs.camptocamp.net/c2cgeoportal/1.6/integrator/security.html

File: c2cgeoportal/lib/caching.py [hunk 9]
```diff
@@ -120,21 +120,25 @@ def set_common_headers(
     else:  # pragma: nocover
         raise "Invalid cache type"
 
+    settings = request.registry.settings.get("headers", {})
```
Comment (sbrunner) [inline c2cgeoportal/lib/caching.py]: Shouldn't you add `.get(service_name, {})` and remove it later?

**Comment by sbrunner** [comment]
You should also add `headers` here https://github.com/camptocamp/c2cgeoportal/blob/add_cors_headers/c2cgeoportal/scaffolds/update/CONST_Makefile_tmpl#L300

File: c2cgeoportal/lib/caching.py [hunk 10]
```diff
@@ -120,21 +120,25 @@ def set_common_headers(
...
+            max_age = settings[service_name]["cache_control_max_age"]
 
         if max_age != 0:
             response.cache_control.max_age = max_age
         else:
             response.cache_control.no_cache = True
 
-    if add_cors:
+    if add_cors and service_name in settings and \
+        "cors" in settings[service_name]:
```
Comment (elemoine) [inline c2cgeoportal/lib/caching.py]: I think we should use `access_control_allow_origin` as the setting name.

**Comment by sbrunner** [comment]
+1 for the doc in security

File: c2cgeoportal/lib/caching.py (hunk 9 above)
Comment (elemoine) [inline c2cgeoportal/lib/caching.py]: I'd rename this to `headers_setting`.

File: c2cgeoportal/lib/caching.py [hunk 11]
```diff
@@ -120,21 +120,25 @@ def set_common_headers(
...
+
     if cache != NO_CACHE:
         max_age = request.registry.settings["default_max_age"]
 
-        settings = request.registry.settings.get("cache_control", {})
-        if service_name in settings and "max_age" in settings[service_name]:
-            max_age = settings[service_name]["max_age"]
+        if service_name in settings and \
+            "cache_control_max_age" in settings[service_name]:
+            max_age = settings[service_name]["cache_control_max_age"]
```
Comment (elemoine) [inline c2cgeoportal/lib/caching.py]: Will this change require notes in the changelog?

File: c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl [hunk 12]
```diff
@@ -115,6 +115,10 @@ vars:
             The GeoMapfish team
         smtp_server: smtp.example.com
 
+    headers:
+        mapserver:
+            cors: *
```
Comment (elemoine) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: Do we really need `*` by default for mapserver?

File: c2cgeoportal/lib/caching.py (hunk 11 above)
Comment (ger-benjamin) [inline c2cgeoportal/lib/caching.py]: Right ! I'll do that

File: c2cgeoportal/lib/caching.py (hunk 10 above)
Comment (elemoine) [inline c2cgeoportal/lib/caching.py]: Does the `add_cors` argument still make sense? It seems silly to add `add_cors=True` everywhere.

Comment (sbrunner) [inline c2cgeoportal/lib/caching.py]: Actually yes, If we want to remove it we should add a `sub_service_name`argument in the function to have a finer granularity.
An than if we do it we should also be able to remove the 
`cache`, `vary` and `content_type` arguments.

File: c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl (hunk 12 above)
Comment (sbrunner) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: Do be able to use it in the `WMSBrowser` of an other site without using the `OGCproxy` who is also a potential hole.

Comment (elemoine) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: > Do be able to use it in the WMSBrowser of an other site

What's the real use-case? Are we talking about API here? Something else?

Comment (sbrunner) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: Use from the Cartoriviera portal with the WMSBrowser the WMS from the Vaud portal without the OGSProxy...

File: c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl [hunk 13]
```diff
@@ -115,6 +115,10 @@ vars:
             The GeoMapfish team
         smtp_server: smtp.example.com
 
+    headers_settings:
```
Comment (elemoine) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: I did not mean to change this to `headers_settings`. Just the variable in the Python code.

File: doc/integrator/security.rst [hunk 14]
```diff
@@ -41,3 +41,17 @@ In the ``viewer.js`` files you should also remove the ``OpenLayers.ProxyHost`` c
...
+By default, only localhost can access c2c's services.
+To permit access to a specific service by an external server, you must set headers CORS (Access-Control-Allow-Origin) in your ``vars_<project>.yaml`` file.
+
+Add or modify the structure like that:
+
+.. code:: yaml
+
+    headers_settings:
+        <service_name>:
+            Access-Control-Allow-Origin: "<domain1>, <domain2>, ..."
```
Comment (elemoine) [inline doc/integrator/security.rst]: Access-Control-Allow-Origin -> access_control_allow_origin

File: c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl (hunk 12 above)
Comment (elemoine) [inline c2cgeoportal/scaffolds/create/vars_+package+.yaml_tmpl]: Use from the Cartoriviera portal with the WMSBrowser the WMS from the Vaud
> portal without the OGSProxy...
>


So this is a specific case. I just think that "*" should not be the default
value.


-- 
Eric Lemoine

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac, Cedex

Tel : +33 (0)4 58 48 20 34
Mail : eric.lemoine@camptocamp.com
http://www.camptocamp.com

File: doc/integrator/security.rst (hunk 14 above)
Comment (sbrunner) [inline doc/integrator/security.rst]: +1

File: c2cgeoportal/lib/caching.py [hunk 15]
```diff
@@ -120,21 +120,24 @@ def set_common_headers(
     else:  # pragma: nocover
         raise "Invalid cache type"
 
+    headers_settings = request.registry.settings.get("headers_settings", {})
+    service_headers_settings = headers_settings.get(service_name)
```
Comment (elemoine) [inline c2cgeoportal/lib/caching.py]: Don't you want to use `.get(service_name, {})` here?

**Comment by sbrunner** [comment]
I think that you should do a rebase for the tests...

**Comment by ger-benjamin** [comment]
(The next week I'll be in holiday (1 week). Please merge or modify if necessary).

**Comment by sbrunner** [comment]
Looks good for me, thanks :-)

**Pull request merged**